codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/DelegatingSyncClientClass.java

```diff
@@ -0,0 +1,180 @@
+                               .addModifiers(Modifier.PRIVATE, Modifier.FINAL)
+                               .build())
+            .addMethod(delegate);
+    }
+
+    @Override
+    protected TypeSpec.Builder createTypeSpec() {
+        return PoetUtils.createClassBuilder(className);
+    }
+
+    @Override
+    protected void addAnnotations(TypeSpec.Builder type) {
+        type.addAnnotation(SdkPublicApi.class);
+    }
+
+    @Override
+    protected void addModifiers(TypeSpec.Builder type) {
+        type.addModifiers(Modifier.ABSTRACT, Modifier.PUBLIC);
+    }
+
+    @Override
+    protected void addFields(TypeSpec.Builder type) {
+    }
+
+    private MethodSpec constructor(ClassName interfaceClass) {
+        return MethodSpec.constructorBuilder()
+                         .addModifiers(Modifier.PUBLIC)
+                         .addParameter(interfaceClass, DELEGATE)
+                         .addStatement("$T.paramNotNull(delegate, \"delegate\")", Validate.class)
+                         .addStatement("this.delegate = delegate")
+                         .build();
+    }
+
+    @Override
+    public ClassName className() {
+        return className;
+    }
+
+    @Override
+    protected MethodSpec.Builder simpleMethodModifier(MethodSpec.Builder builder) {
+        return builder.addAnnotation(Override.class);
+    }
+
+    @Override
+    protected MethodSpec.Builder operationBody(MethodSpec.Builder builder, OperationModel opModel) {
+        builder.addModifiers(Modifier.PUBLIC)
+               .addAnnotation(Override.class);
+
+        if (opModel.hasStreamingInput() || opModel.hasStreamingOutput()) {
+            String variableName = opModel.hasStreamingInput() ? SYNC_STREAMING_INPUT_PARAM : SYNC_STREAMING_OUTPUT_PARAM;
+            return builder.addStatement("return delegate.$N($N, $N)",
+                                        opModel.getMethodName(),
+                                        opModel.getInput().getVariableName(),
+                                        variableName);
+        }
+
+        if (opModel.hasEventStreamInput() && opModel.hasEventStreamOutput()) {
+            return builder.addStatement("return delegate.$N($N, $N, $N)",
+                                        opModel.getMethodName(),
+                                        opModel.getInput().getVariableName(),
+                                        EVENT_PUBLISHER_PARAM_NAME,
+                                        SYNC_STREAMING_OUTPUT_PARAM);
+        }
+
+        if (opModel.hasEventStreamInput() || opModel.hasEventStreamOutput()) {
+            String variableName = opModel.hasEventStreamInput() ? EVENT_PUBLISHER_PARAM_NAME : EVENT_RESPONSE_HANDLER_PARAM_NAME;
+            return builder.addStatement("return delegate.$N($N, $N)",
+                                        opModel.getMethodName(),
+                                        opModel.getInput().getVariableName(),
+                                        variableName);
+        }
```

Comment:
nit: any way to just copy the delegate method name and parameters from the incoming MethodSpec.Builder, since the caller already has it figured out?

Something like:
```java
builder.addStatement("return delegate.$N($L)", 
                     opModel.getMethodName(), 
                     builder.parameters.stream().map(p -> p.name).collect(joining(", "));
```

(I haven't tested this, I could be way off).

Reply:
that worked seamlessly! :)